internal/backend/sema/backend.go

```diff
@@ -0,0 +1,99 @@
+package sema
+
+import (
+	"context"
+	"io"
+
+	"github.com/cenkalti/backoff/v4"
+	"github.com/restic/restic/internal/debug"
+	"github.com/restic/restic/internal/errors"
+	"github.com/restic/restic/internal/restic"
+)
+
+// make sure that SemaphoreBackend implements restic.Backend
+var _ restic.Backend = &SemaphoreBackend{}
+
+// SemaphoreBackend limits the number of concurrent operations.
+type SemaphoreBackend struct {
+	restic.Backend
+	sem semaphore
+}
+
+// New returns a new backend that saves all data in a map in memory.
+func New(be restic.Backend) *SemaphoreBackend {
+	sem, err := newSemaphore(be.Connections())
+	if err != nil {
+		panic(err)
+	}
+
+	return &SemaphoreBackend{
+		Backend: be,
+		sem:     sem,
+	}
+}
+
+func (be *SemaphoreBackend) validateAndGetToken(h restic.Handle) (func(), error) {
```

Comment:
I know this is supposed to prevent repetition, but I don't find "validate Handle and grab token from semaphore" an obvious action. If we put the debug.Log statement inside semaphore.GetToken, inlining this at the call sites wouldn't produce much longer code.

Reply:
I've inlined the method.